Inverted Mode

cassidoxa opened the pull request:
Notes:

- Added inverted ROM changes or applied ER ROM fixes where appropriate
    -Includes text changes for intro and spawn points
- Added new module InvertedRegions.py which reflects inverted overworld changes
- Added mandatory and default connections for inverted in EntranceShuffle.py
- Added general rules, new bunny rules, and new big bomb rules for inverted in Rules.py
    -Changed some dungeon logic for single-entrance dungeons like IP and MM to more closely resemble VT logic, removing rules for doors but compensating with rules for chests
- Link's House is freely shuffled, Dark Sanc is shuffled in West Dark World to strike balance between non-entrance inverted and non-inverted ER
- Old Man spawn point moved to door where his cave exits as in non-entrance inverted, can exit on W DM or W DDM because of portal on W DDM
    -Spawn menu text changes to reflect where this door is
- In shuffles where Aga Tower can be on HC Ledge, Aga Tower is shuffled first or early with one ledge door becoming a must-exit if it's on the ledge to ensure access
- Added pseudo-cave "The Sky" to represent flute connections. Flute works in DW and connects potentially-unreachable regions
- Added exits in LW regions that connect to HC Ledge after Aga is killed

Known Issues:

- Filler occasionally can't place TR keys in some configurations. Most often when Big Key is in the front three chests and a small key is placed in Chain Chomps and Big Chest.
    -Mostly affects dungeonsfull shuffle, comes up rarely in full+
    -Still working on this, can send another patch when it's figured out

cassidoxa committed:
Add inverted to CLI options

cassidoxa committed:
Initial inverted copy paste job from VT Rom.php

cassidoxa committed:
More ROM stuff
Set dark sanc to be static for now
Started on inverted rule set

cassidoxa committed:
Fix function that writes multiple int16 as bytes

cassidoxa committed:
Basic inverted framework
Need to make regions, rules, and adjust shuffle algo

cassidoxa committed:
Inverted regions

cassidoxa committed:
LW Mirror spots

cassidoxa committed:
Mandatory connections

cassidoxa committed:
Flute connections
More mirror stuff

cassidoxa committed:
Regions and default/mandatory connections mostly done

cassidoxa committed:
Inverted rules and bunny rules
Changes in Main, BaseClasses, etc to account for inverted
Should be able to generate inverted vanilla shuffle seeds

cassidoxa committed:
Fix Bumper Cave Ledge connections
Randomizer should no longer place progression on bumper ledge without
requiring moon pearl, mirror, LW access

cassidoxa committed:
Change spawn menu for inverted

cassidoxa committed:
Started implementing simple shuffle for inverted

cassidoxa committed:
Working shuffle implementations up to simple shuffle

cassidoxa committed:
Added can_flute function to base classes

cassidoxa committed:
Inverted flute spots
Fixed post Aga 1 HC Ledge exit connections

cassidoxa committed:
Simplified inverted flute implementation

cassidoxa committed:
Fixed implementation of dungeons simple shuffle
Added some rules requiring Pearl to pick up bushes in LW

cassidoxa committed:
Added new mirror spots
Changed MM key logic, removed west door logic and added reqs to chests

cassidoxa committed:
Updated inverted rules
added can_melt_things function to base classes
adjusted some dungeon logic to be more like VT logic

cassidoxa committed:
Removed some always-accessible entrances in Rules.py

cassidoxa committed:
Misc tweaking of rules and connections

cassidoxa committed:
Fixed connections related to IP entrance and Dark Hylia teleporter
Fixed can_melt_things function

cassidoxa committed:
Finished inverted simple shuffle
Dark Sanc can now be moved

cassidoxa committed:
Working implementations up to crossed

cassidoxa committed:
More polished working implementations up to crossed shuffle
Links House is shuffled freely, Dark Sanc shuffled in West Dark World

cassidoxa committed:
Swamp and powder patches for inverted
added checks for vanilla GT and Ganon hole

cassidoxa committed:
Fix GT/AT mixup

cassidoxa committed:
Fix DP key logic

cassidoxa committed:
Added LW DM doors as old man doors in inverted crossed

cassidoxa committed:
Working implementation of insanity shuffle

cassidoxa committed:
Refactor shuffled links house/dark sanc into neater functions
Added "shuffle_links_house" and "shuffle_dark_sanc" boolean attributes to world
Should make it trivial to shuffle links house in non-inverted
Cleaned up print debugging

cassidoxa committed:
Set inverted big bomb rules
Add flute exits to Dark Grassy Lawn and Hammer Peg Area
Clean up Main.py

cassidoxa committed:
Fix Old Man Cave West Exit in full

cassidoxa committed:
Add inverted to Gui.py

cassidoxa committed:
Old Man Logic

cassidoxa committed:
Old Man spawn menu text and logic

cassidoxa committed:
Non vanilla inverted Ganon hole

cassidoxa committed:
Fix Inverted Pyramid Entrance overworld door id

cassidoxa committed:
Text

cassidoxa committed:
Houlihan connections boilerplate and minor fixes

cassidoxa committed:
Function for writing groups of 16 bit bytes for JSON ROM

cassidoxa committed:
Fix DP location requirements and some LW bush drop requirements
Quick fix for potion shop: now requires pearl

cassidoxa committed:
Added Northeast Light World and River regions
Better represents connections in this area
Should prevent issues with access to mushroom turn in

cassidoxa committed:
Northeast Light World Region

cassidoxa committed:
Cleanup of no glitches rules and inverted trock entrance/exit fixes

cassidoxa committed:
No Logic
General Tidying up

cassidoxa committed:
Update readme

cassidoxa committed:
Added connections from LW regions to HC Ledge after Aga 1 has been
killed

cassidoxa committed:
Add Pearl requirement for HC Drop

cassidoxa committed:
Fix patcher overwriting inverted pyramid entrance when shuffled

cassidoxa committed:
Accidental commit

cassidoxa committed:
Fixed extra inverted Ganon hole on HC ledge sending Link to Houlihan
instead of target

cassidoxa committed:
Renamed 'Dark Lake Hylia Ledge' exit to 'Dark Lake Hylia Ledge Pier' in
inverted, added flippers requirement

cassidoxa committed:
Full shuffle can put Old Man exit on LW or DW West DM
Old Man Cave entrances will be in same world

cassidoxa committed:
Inverted Dark Sanc exit

cassidoxa committed:
Dark Sanc

cassidoxa committed:
Changed Pyramid Hole to Inverted Pyramid Hole in default inverted
connections

cassidoxa committed:
Adjust aga shuffle in inverted dungeonsfull, can't be on HC ledge
avoids unfillable configurations when TR big chest door is on HC or
desert ledge

cassidoxa committed:
Old Man Cave can be in light or dark world in inverted restricted
shuffle
Cave will remain in same world

cassidoxa commented:
The dungeonsfull issue seems to have been solved by placing aga tower off HC Ledge. I think the issue was cropping up when the TR Big Chest door was placed at a must exit entrance on Desert or HC Ledge. Another solution might be to ensure that this exit doesn't connect to Desert Ledge or as the only must-exit on HC Ledge.

cassidoxa committed:
Maze Race Req

cassidoxa committed:
Maze Race

cassidoxa committed:
Dark Desert Ledge

cassidoxa committed:
Move inverted ROM stuff into its own function

cassidoxa commented:
I think most of the major issues have been ironed out. The TR key issue has been happening for me on the master branch as well so it seems its not unique to inverted. If you're interested in accepting the PR, feel free to let me know if there's any changes you want to see or anything like that and I'd be happy to make them.

TheExcalabur commented:
Bug: The dark chapel guy still sells you a hint and then does the heal.  This hint prompt should be transferred to sanctuary? (I haven't check what happens without hints yet)
Design query: What's up with the mountain load point? This isn't how either inverted or entrance rando works?  Why aren't we just loading at Old Man House like we usually do?
Design suggestion: Add the Houlihan room and the pyramid crack to the set of rooms to be shuffled.     This could provide cool out-of-logic routing decisions.   Pyramid crack has to be added to the must-exit list.
I haven't yet checked the connection logic in detail to make sure it's bug-free.  This is on my to-do list.

(Good job--glad someone finally put the time in to do this.)

cassidoxa commented:
The way the mountain spawn is handled in inverted, it puts you at the door where the Old Man Cave exits. For the most part I copied over the inverted stuff in Rom.php from the VT rando so that's how it ended up. This does probably need a better solution. Could keep it at the Old Man House or maybe at Dark Death Mountain Fairy? There's not necessarily a compelling reason to change it from Old Man House in ER I suppose. I'd need to adjust some of the inverted shuffles since I put it in with the regular caves instead of handling it separately as the regular shuffles do.

Moving the Dark Sanc hint to the Sanctuary is probably a good idea. I'm sort of familiar with how the text is handled, but I'm not sure I'd be able to remove healing from Dark Sanc and add a text box to the guy in Sanctuary. I'll look into it but hopefully someone can help out with that bit. And yeah, the pyramid crack is probably better as a must-exit in ER.

Thanks.

TheExcalabur commented:
I think just standard Old Man House is the right answer for the Old Man spawn--there's no compelling reason to change it.  In non-full shuffles some care might be taken to make sure that it's on DM/DDM.  

Unfortunately I can't help with the text stuff without learning it in the first place.  

Pyramid Crack is actually a mirrorable location in inverted--you can get there from on top of Hyrule castle without having to do anything weird.  

Stuff from code review:
Link and Dark Chapel should be able to connect to must-exits where appropriate—Link’s house to any of them, Dark Chapel none for now.  Can be done just by selecting the entrance from the list of must_exits + entrances, as long as care is taken.

Dark chapel in ‘west dark world’ could include 
“Bumper Cave (Bottom)”
“Bumper Cave (Top)”
“Skull Woods Final Section”
“Thieves Town”
for shuffles that allow multi exit mixing or dungeon mixing.  I think you've already gotten TT where appropriate.  SW back might be a bit weird, but seems fine--if SW is arranged wrong item logic should be able to fix it.  

We should randomly choose if Desert back or Desert left is the must-exit location, as they're logically distinct due to the rocks.  (It'd be nice to choose which side of TR bridge was must-exit too, but it doesn't make a whole ton of difference.)

cassidoxa committed:
Change Old Man spawn back to Old Man Cave in ROM

cassidoxa commented:
So I've been experimenting with that Pyramid Door. I've figured out how to enter Houlihan from an overworld door, but inverted moves that door to the new HC castle entrance. While I'd like to include it in the shuffle, it looks like it would require some extra rom hacking to add a brand new OW door. I could maybe try it using some empty space but there's the possibility that one of the devs may want to use that space in a future base ROM. And I'm not sure if it's even possible with my current knowledge so I'm going to leave it as it is for now and maybe revisit.

I changed the spawn point back to Old Man Cave in the ROM. So I'll adjust the shuffles to account for this as well. Additionally I'll fix the Dark Sanc and Link's House shuffles and randomize the  Desert must-exit per your suggestions.I'll try to push an update today or tomorrow. 

Thieves' Town is already a possible Dark Sanc location, but it's added in during some shuffles and not in the base variable since it could end up there in simple dungeon shuffles if it was. It's similar to how Tower of Hera is added as an Old Man Entrance in non-inverted. As for Skull Woods, it seems like you can't walk out of there without fire rod-ing the thing in front of the door. Is that how it usually works? My concern with Dark Sanc being there was not having access to West Dark World and Link's House potentially being in a bad spot.

cassidoxa committed:
Logic changes to account for reverted Old Man spawn point
Added possible Dark Sanc and Link's House doors
Desert Ledge must-exit is randomized

TheExcalabur commented:
Inverted has one more entrance than the standard rom, which is why they de-linked Houlihan.  I'm surprised something weird needs to be done here, but SNES roms continue to surprise.  Does the crack not two-way connect to Houlihan in the inverted "base" rom?  I'll clone the repo and have a fiddle.  

Do you have a map of the (logical) regions? Trying to reconstruct one from the code is somewhat difficult.

The back of skull woods usually works like other one-way exits--you get kicked out just far enough out that you can move.  Not sure why that would be different.  (Now I want to check that to make sure I'm not telling lies.)  My thought with using SW back as a possible location is that--very occasionally--you wouldn't have early access to WDW; it would also make S&Q slightly less appealing.  It wouldn't happen that often, and I think there's enough resiliency in the system that it wouldn't be too frustrating.

cassidoxa commented:
The connection from Houlihan is only one-way. I'm not sure about the base ROM but my understanding is that it "moves" the overworld door that was at the pyramid to the one by Hyrule castle. So the game sort of re-uses some data for the former treats the old door like a hard wall. So I think an additional door would require another entry into the exit table specifying the location and camera position and all that. But I'm not sure if there's extra room in the ROM where the current exit table is.

I'll add in the Skull Woods Final Section door as a sanctuary door. I don't have a visual map. I could maybe throw one together in GIMP. I think the main differences are Northeast Light World is its own region due to the bushes and water which the bunny can't cross, IP is just a part of Dark Lake Hylia ( I think I added flippers requirements to the piers so the randomizer doesn't think you can use them if you can exit the IP door without them), DDM is split into Dark Death Mountain and Dark Death Mountain (East Bottom) because of the stairs, and Dark Desert Ledge is its own region since you can only reach the portal with the flute.

Edit: I tested the SW Dark Sanc exit and wasn't able to move out of the door. I guess I'll leave it in for now pending any objections. I'm fine with the 1/11 chance that the player won't have guaranteed access to W DW. I'll have to make sure the logic doesn't assume you can get to West Dark World that way without the fire rod though.

cassidoxa committed:
Add addition Dark Sanc doors
Fix ugly print debugging

cassidoxa committed:
Put Dark Sanc in West Dark World in insanity shuffle

cassidoxa committed:
Connect Links House two-way in insanity

TheExcalabur commented:
Bug:  Bunny link traversal in the LW isn't quite right.   I rolled a seed (easy-inverted-crystals_crossed-balanced-keysanity_243853716) on the version that was up a few hours ago that put link's house at the overgrown house in Kak.   The logic assumed that I could get from there to the 'Light World' region, which you cannot, in sphere 1.  I'm not sure offhand which other entrances are bush-blocked/liftable-blocked in inverted in such a way that bunny traversal fails, but my guess is that some regions are missing and/or connections don't include the right set of items.  

Bug: the route log seems to also think that I could open the map chest in HC and the chests in paradox having traversed to minimoldorm without a stop in the dark world, so that's another bug.  You can DBR to get the HC chest, but that should be out of logic.  I wonder if something's not quite right with the save point being in the bunny world?

Spoiler log attached

[ER_noglitches_easy-inverted-crystals_crossed-balanced-keysanity_243853716_Spoiler.txt](https://github.com/AmazingAmpharos/ALttPEntranceRandomizer/files/3332078/ER_noglitches_easy-inverted-crystals_crossed-balanced-keysanity_243853716_Spoiler.txt)

cassidoxa commented:
Thanks. I added a pearl requirement to that house with the bushes, but it may actually need a new region and connections instead.

Edit: For Hyrule Castle and Paradox I think the logic expects you to exit the doors in those locations which lead to the Dark World then re-enter as Link. I've also added a pearl requirement to mini moldorm cave. I thought I'd added it already.

cassidoxa committed:
New region, connections, and rules for LW Bush Covered House Lawn

cassidoxa committed:
Update Old Man spawn point logic for vanilla shuffle

cassidoxa committed:
Pearl requirement for mini moldorm cave

cassidoxa committed:
Typo

cassidoxa committed:
Pearl requirement for LW Hype Cave

cassidoxa committed:
Revised Northeast Light World regions and logic
Added new region and logic for Graveyard Cave

cassidoxa committed:
Add LW Bomb Hut region, connections, logic

cassidoxa commented:
Here's maps of the logical regions, with some of the inverted changes notated in blue as well https://imgur.com/a/FKRKFQF

cassidoxa committed:
Bomb Hut Bushes

cassidoxa committed:
can_flute and good bee access

cassidoxa commented:
To Kevin:

Let me know if there's any changes you'd like to see. I made new keys for Inverted Link's House, Bomb Shop, Aga Tower, and GT in `door_addresses` and `exit_ids` when the vanilla ones could probably be used. The Link's House shuffle was a design choice based on feedback from some cross keys people. That could be changed back to inverted vanilla or refined further. I ended up writing some bytes to the `StartingAreaExit` table for Link's House when I was trying to figure out shuffling that, not sure if it was necessary. Ideally the old Pyramid Entrance could be re-used but that's probably beyond my abilities. I also put the inverted ROM stuff into a new function which only writes some bytes when the shuffle is vanilla. For example the Old Man spawn is on DDM in vanilla but at Old Man House in inverted and the logic accounts for this. 

The flute also uses a new fake cave. This may be buggy, not 100% sure but sometimes it will hang when generating seeds and it seems to be related to the flute. Maybe a recursion issue or something. ~~Lastly, I haven't tested the new `write_int16_bytes` method for the JsonRom object yet.~~

Edit: Also I noticed in the v31 doc that Old Man House overworld doors will be moved to West DDM, so that would need some changes in InvertedRegions.py and maybe the door addresses/exit ids as well.

cassidoxa committed:
Cleanup, big bomb rules

cassidoxa committed:
HC Secret Passage Bushes

KevinCathcart commented on BaseClasses.py:
```diff
@@ -483,13 +484,22 @@ def can_shoot_arrows(self):
+                self.has('Bug Catching Net') and
+                (self.has_Boots() or (self.has_sword() and self.has('Quake'))) and
+                cave.can_reach(self) and
+                (cave.is_light_world or self.has_Pearl())
```
Perhaps we can add an `is_not_bunny(self, region)` helper, that checks for pearl or light/dark world as applicable. It would simplify checks like this.

KevinCathcart commented on Dungeons.py:
```diff
@@ -27,13 +27,42 @@ def make_dungeon(name, default_boss, dungeon_regions, big_key, small_keys, dunge
+    
+    world.dungeons = [ES, EP, DP, ToH, AT, PoD, TT, SW, SP, IP, MM, TR, GT]
+
+def create_inverted_dungeons(world):
```
I'm not the biggest fan of all this duplication. If we can come up with a good way to avoid it it would be nice. It is not obvious how best to do that though. See if you can come up with some good way to avoid it. This is not a blocker though.

KevinCathcart commented on BaseClasses.py:
```diff
@@ -512,6 +522,18 @@ def has_Pearl(self):
+        if self.world.mode != 'inverted':
+            return self.has('Ocarina') and self.world.get_region('Light World').can_reach
+        elif self.world.mode == 'inverted':
+            return self.has('Ocarina') and self.has_Pearl() and self.world.get_region('Light World').can_reach
```
This could also use that helper, to remove the if.

KevinCathcart commented on BaseClasses.py:
```diff
@@ -910,7 +932,11 @@ def parse_data(self):
+        # inverted_lw_locations = [loc for loc in self.world.get_locations() if loc not in listed_locations and loc.parent_region and loc.parent_region.type == RegionType.InvertedLightWorld]
+        # self.locations['Inverted Light World'] = OrderedDict([(str(location), str(location.item) if location.item is not None else 'Nothing') for location in lw_locations])
+        # listed_locations.update(lw_locations)
+        
```
Let's not add new commented out code.

KevinCathcart commented on EntranceShuffle.py:
```diff
@@ -2187,12 +3689,14 @@ def tuplize_lists_in_list(ls):
                   'Hookshot Cave': (0x39, (0x003c, 0x45, 0x04da, 0x00a3, 0x0cd6, 0x0107, 0x0d48, 0x0112, 0x0d53, 0x0b, 0xfa, 0x0000, 0x0000)),
                   'Hookshot Cave Back Entrance': (0x3A, (0x002c, 0x45, 0x004c, 0x0000, 0x0c56, 0x0038, 0x0cc8, 0x006f, 0x0cd3, 0x00, 0x0a, 0x0000, 0x0000)),
                   'Ganons Tower': (0x36, (0x000c, 0x43, 0x0052, 0x0000, 0x0884, 0x0028, 0x08f8, 0x006f, 0x0903, 0x00, 0xfc, 0x0000, 0x0000)),
+                  'Inverted Ganons Tower': (0x23, (0x00e0, 0x1b, 0x0032, 0x0600, 0x0784, 0x0634, 0x07f8, 0x066d, 0x0803, 0x00, 0x0a, 0x0000, 0x82be)),
```
I'm torn about the duplicate door_addresses. 
However to the extent that we have them, I'd prefer to see them next to the row they duplicate.

Obviously we cannot completely get rid of them, since Inverted Pyramid Hole has the extra address.

cassidoxa replied:
I made duplicate door_addresses and exit_ids early on to make it a bit easier for me to follow , and I wasn't certain if the door index or room id needed to be changed for inverted or not. If that's not the case then the only reason for duplicates would maybe be for the spoiler logs.

KevinCathcart commented on EntranceShuffle.py:
```diff
@@ -2275,7 +3781,8 @@ def tuplize_lists_in_list(ls):
 exit_ids = {'Links House Exit': (0x01, 0x00),
-            'Chris Houlihan Room Exit': (None, 0x3D),
+            'Inverted Big Bomb Shop': 0x53,
+            'Chris Houlihan Room Exit': (0x82, 0x3D),
```
Technically, the Houlihan exit does not have an entry in the entrance table.

There is an entry for the Houlihan hole target location (which is indeed 0x82), but that is a separate entrance, which really would belong near the bottom on exit_ids. (which is misleadingly named, since it has both entrance and exit ids.

Was there a reason you needed to specify a non-None value here?

cassidoxa replied:
In this case I was trying to add the non-inverted Pyramid Entrance back in to the shuffle and needed an extra cave.

KevinCathcart replied:
Yeah, since we pyramid exit to the light world, if we were to have something shuffled there, we would need an extra cave. If i were doing that though, I would likely consider adding say an extra instance of the mutliply used fairy fountain or dark world shop. It would be a bit less jarring than entering a cave and finding yourself in the middle of the Houlihan room.

As far as adding that as an extra entrance we would need to add an entry to the overworld doors table to handle it. Technically there is one entry in that that that is unused (it is a duplicate) so it would be possible. But is it worth it for just adding a fairly useless location, since it is not reachable, except by exiting from it?

cassidoxa replied:
The old Pyramid Entrance can be mirrored to from HC ledge in inverted, on the lower left near where it widens out. But yeah, it's not super important to have especially since access to HC ledge is limited as well.

KevinCathcart commented on EntranceShuffle.py:
```diff
@@ -2286,6 +3793,7 @@ def tuplize_lists_in_list(ls):
             'Hyrule Castle Exit (West)': (0x03, 0x02),
             'Hyrule Castle Exit (East)': (0x05, 0x04),
             'Agahnims Tower Exit': (0x24, 0x25),
+            'Inverted Ganons Tower Exit': (0x37, 0x38),
```
Once again, If we are going to have duplicates, let's keep these duplicates near the entry they duplicate.

KevinCathcart commented on Rules.py:
```diff
@@ -169,7 +193,7 @@ def global_rules(world):
     set_rule(world.get_entrance('Dark Lake Hylia Ledge Fairy'), lambda state: state.has_Pearl()) # bomb required
     set_rule(world.get_entrance('Dark Lake Hylia Ledge Spike Cave'), lambda state: state.can_lift_rocks() and state.has_Pearl())
-    set_rule(world.get_entrance('Dark Lake Hylia Teleporter'), lambda state: state.has_Pearl() and (state.has('Hammer') or state.can_lift_rocks()))  # Fake Flippers
+    set_rule(world.get_entrance('Dark Lake Hylia Teleporter'), lambda state: state.has('Flippers') and state.has_Pearl() and (state.has('Hammer') or state.can_lift_rocks()))  # Fake Flippers
```
Hmm... this is the whirlpool warp from north east dark world to lake hylia. Fake flippers I'm fairly certain fake flippers are possible here (The flippers requirement for glitchless gets added in `no_glitches_rules`.).

However it *is* in the wrong region right now, and if properly put in North East Dark World, this would be simplfied to has_pearl().

KevinCathcart commented on Rules.py:
```diff
@@ -200,7 +224,6 @@ def global_rules(world):
 
     set_rule(world.get_entrance('Desert Ledge Mirror Spot'), lambda state: state.has_Mirror())
     set_rule(world.get_entrance('Desert Palace Stairs Mirror Spot'), lambda state: state.has_Mirror())
-    set_rule(world.get_entrance('Desert Palace Entrance (North) Mirror Spot'), lambda state: state.has_Mirror())
```
Why was this removed in non-inverted? I'm pretty sure we don't want people to be able to mirror from dark world to light world without having a mirror!

cassidoxa replied:
Thanks for pointing this out, this was just me being careless with my editor's search.

KevinCathcart commented on Rules.py:
```diff
@@ -315,12 +336,13 @@ def global_rules(world):
-    set_rule(world.get_location('Misery Mire - Big Key Chest'), lambda state: state.has_fire_source())
+    set_rule(world.get_location('Misery Mire - Big Key Chest'), lambda state: state.has_fire_source() and 
+                                                                    (item_name(state, 'Misery Mire - Big Key Chest') == 'Big Key (Misery Mire)' and state.has('Small Key (Misery Mire', 2)) or state.has('Small Key (Misery Mire', 3))
     set_rule(world.get_entrance('Misery Mire (Vitreous)'), lambda state: state.has('Cane of Somaria'))
```
I'll need to think about these Mire changes.

cassidoxa replied:
I ported over the VT logic for Mire and Ice Palace, so that's where this came from. My thinking was that as single entrance dungeons it would be a bit simpler to only consider the item locations and the key requirements being moved from doors to chests should prevent unwinnable games.

KevinCathcart commented on Rom.py:
```diff
@@ -63,13 +67,21 @@ def write_byte(self, address, value):
     def write_int16(self, address, value):
         self.write_bytes(address, int16_as_bytes(value))
 
+    def write_int16_bytes(self, startaddress, values):
```
Umm... what is an int16 byte? Consider renaming as `write_int16s`.

KevinCathcart commented on Rom.py:
```diff
@@ -63,13 +67,21 @@ def write_byte(self, address, value):
     def write_int32(self, address, value):
         self.write_bytes(address, int32_as_bytes(value))
 
+    def write_int32_bytes(self, startaddress, values):
```
Same.

KevinCathcart commented on Rom.py:
```diff
@@ -340,7 +352,7 @@ def patch_rom(world, rom, hashtable, beep='normal', color='red', sprite=None):
                     elif room_id == 0x00d6 and world.fix_trock_exit:
                         rom.write_int16(0x15DB5 + 2 * offset, 0x0134)
-                    elif room_id == 0x000c and world.fix_gtower_exit: # fix ganons tower exit point
+                    elif room_id == 0x000c and world.fix_gtower_exit and world.mode != 'inverted': # fix ganons tower exit point
```
Surely this could be handled by not setting world.fix_gtower_exit for inverted mode, and thus omitting that check here?

KevinCathcart commented on Rom.py:
```diff
@@ -735,12 +756,13 @@ def patch_rom(world, rom, hashtable, beep='normal', color='red', sprite=None):
+        rom.write_byte(0x3A943, 0xD0) # Mirror: Normal  (D0=Dark to Light, F0=light to dark, 42 = both)
+        rom.write_byte(0x3A96D, 0xF0) # Residual Portal: Normal  (F0= Light Side, D0=Dark Side, 42 = both (Darth Vader))
+        rom.write_byte(0x3A9A7, 0xD0) # Residual Portal: Normal  (D0= Light Side, F0=Dark Side, 42 = both (Darth Vader))
 
```
I'd prefer to pull out the corresponding inverted writes to these variables, and put them here.

KevinCathcart commented on Rom.py:
```diff
@@ -1204,6 +1238,232 @@ def write_strings(rom, world):
 
+
+def set_inverted_mode(world, rom):
+    rom.write_byte(snes_to_pc(0x30804A), 0x01)  # inverted main toggle
```
Redundant with line 759

KevinCathcart reviewed (changes requested):
I have not reviewed the inverted regions or logic.
But I have looked over most of the rest, and have a few comments/concerns.

cassidoxa committed:
First batch of changes based on Kevin's feedback

create_dungeons now creates inverted dungeons, removed
create_inverted_dungeons
Fixed a couple rules, added Desert North mirror spot back in
non-inverted
Cleaned up ROM stuff, renamed write_int16/32_bytes

cassidoxa commented:
Thanks a lot for the feedback and suggestions. I've tackled most of the low-hanging fruit here. I'll leave it up to you whether the duplicate door_addresses and exit_ids remain. I'll admit that's probably something I should have addressed before now.

cassidoxa committed:
is_not_bunny helper function

cassidoxa committed:
SImplyify is_not_bunny

KevinCathcart commented on InvertedRegions.py:
```diff
@@ -161,9 +161,9 @@ def create_inverted_regions(world):
+        create_dw_region('East Dark World', ['Pyramid'], ['Pyramid Fairy', 'South Dark World Bridge', 'Palace of Darkness', 'Dark Lake Hylia Drop (East)', 
                                                           'Dark Lake Hylia Fairy', 'Palace of Darkness Hint', 'East Dark World Hint', 'Northeast Dark World Broken Bridge Pass', 'East Dark World Teleporter', 'EDW Flute']),
-        create_dw_region('Northeast Dark World', ['Catfish'], ['West Dark World Gap', 'Dark World Potion Shop', 'East Dark World Broken Bridge Pass', 'NEDW Flute']),
+        create_dw_region('Northeast Dark World', ['Catfish'], ['West Dark World Gap', 'Dark World Potion Shop', 'East Dark World Broken Bridge Pass', 'NEDW Flute', 'Dark Lake Hylia Teleporter']),
```
Could you also make this change to Regions.py?

cassidoxa committed:
Move Dark Lake Hylia Teleporter to NE Dark World

KevinCathcart commented:
Also, please go ahead and mark any resolved comment conversations as resolved. I would do it myself, but i actually don't have push access to AA's repo. Leave any with open questions open obviously.

When I'm ready to pull these changes in, I'll be pulling it into my repo (which is upstream of AA's), but I don't see any reason not to continue the discussion in this pull request thread. 

Generally, I am liking what I am seeing here, and do intend to merge this in at some point. That could be a few more days, and there will probably be a bit more work needed on your part, but I'll get into that a bit later.

cassidoxa commented:
Awesome, thanks again for the review.

cassidoxa committed:
Cleanup, delete unused variables and commented code
Aga Tower is handled consistently across inverted shuffles
Old Man House can be in either world in full inverted shuffle
Adjusted connect_mandatory_exits to account for randomized dp must-exit
Fix my breakage of non-inverted simple shuffle

cassidoxa committed:
Fix Spike Cave handling in inverted simple shuffle

cassidoxa committed:
Fix inverted simple shuffle
Put Spike Cave back into inverted bomb shop doors

KevinCathcart commented:
OK, so having thought things over, I'd actually prefer not to have the dungeon changes. VT's Ice palace logic is actually not completely compatible with monotonicity requirement that ER imposes (this is due to the ternary operator used in that logic).

Other than that I *would* be ready ready to merge this in, except for one thing.

Over  on the Dev branch of my repository, I've pulled in the core set of changes from Bonta0's multiworld fork, in preparation of eventually supporting the official ALTTPR multiworld system.

This makes a lot of changes to support having player numbers all over the place, and it will cause a huge number of merge conflicts with this patch. Also in several spots where no merge conflict would occur, there would still need to be required changes to work properly.

If you could merge in (or rebase on top of) my Dev Branch, and open a pull requerst over in my repository for it, I would greatly appreciate it. (It would likely take me a lot longer to do that myself, given my limited time availability at the moment.)

cassidoxa commented:
Ok, I'll revert the dungeon changes and work on the new PR. I did try to merge inverted with bonta's multiworld a few weeks ago and only had an issue with the multiworld balancing, which seemed to loop infinitely. So that's the only thing I may need some help with, but I'll ask bonta about it if I can't figure it out before submitting a PR. 

I think what I'll do is pull your dev branch into a new branch of this repo, merge in this branch manually, then submit a PR with that if that's acceptable. I'm still kind of new to git and that seems like the easiest way for me to do it.

KevinCathcart commented:
Yeah, that would be fine. With the sorts of conflicts that would occur, the automatic merge is not likely to be very helpful, so doing it manually is likely to be easier anyway. 

And don't worry too much about the balancing part. I already know it is somewhat buggy. (It has managed to generate seeds that cannot be beaten, according to ER's own internal Logic checks, at least when only 2 worlds are generated. I've not seen that happen with 3+ worlds). I've not actually looked at that part very closely yet.

If you can get a merged version that works for single world, I'd be more than happy to merge that in pending a determination of why the balancing fails.

KevinCathcart commented:
Also if you have any questions about what's different in my Dev branch vs Bonta0's or anything like that, please don't hesitate to DM me on Discord. You can find me in the main ALTTPR Discord, and and I'm sure you can figure out my name there. :-D
